### PR TITLE
ZIR-222: Add keccak lift predicate

### PR DESCRIPTION
**Pull request by shkoo**
* move calculator to its own subdirectory
* Add `TapsAttr` module attribute as authorative source for the list of taps
* update wrap_zirgen predicate to work with TapsAttr
* Remove unused MakeVerifyTaps
* Fix calculator codegen so we can mostly run end-to-end again
* Get calculator predicate working in test, fixed bugs
* Generate keccak predicate
* Fix problem with MulOp folding, where it emits a constant that might not be the same as the return type

Still todo:

* Hook keccak up to zkvm
* Test keccak lift predicate
* Clean up calculator example

### CODE REVIEW

File: zirgen/circuit/verify/wrap_zirgen.cpp [hunk 1]
```diff
@@ -231,18 +245,17 @@ Val CircuitInterfaceZirgen::compute_poly(llvm::ArrayRef<Val> u,
...
-std::unique_ptr<CircuitInterface>
-getInterfaceZirgen(mlir::MLIRContext* ctx, mlir::StringRef filename, ProtocolInfo protocolInfo) {
+std::unique_ptr<CircuitInterface> getInterfaceZirgen(mlir::MLIRContext* ctx,
+                                                     mlir::StringRef filename) {
   ParserConfig config(ctx);
   OwningOpRef<ModuleOp> mod = parseSourceFile<ModuleOp>(filename, config);
   if (!mod)
     return nullptr;
-  return std::make_unique<CircuitInterfaceZirgen>(std::move(mod), protocolInfo);
+  return std::make_unique<CircuitInterfaceZirgen>(std::move(mod));
```
Comment (jacobdweightman) [inline zirgen/circuit/verify/wrap_zirgen.cpp:258]: I'm rather confused about what's going on here, could you describe what the purpose of this file is?

**Review by jacobdweightman — APPROVED** [review]

File: zirgen/Dialect/Zll/IR/Attrs.td [hunk 2]
```diff
@@ -26,22 +26,32 @@ class ZllAttr<string name, string attrMnemonic, list<Trait> traits = []>
 def Tap : ZllAttr<"Tap", "tap"> {
   let summary = "A Tap, which specifies a register and a back.";
   let parameters = (ins
-      "uint32_t" : $regGroupId,
-      "uint32_t" : $offset,
-      "uint32_t" : $back);
+      "unsigned" : $regGroupId,
+      "unsigned" : $offset,
+      "unsigned" : $back);
```
[the remark this answers is not in the document]
Comment (shkoo) [reply zirgen/Dialect/Zll/IR/Attrs.td:31]: Not necessary, I'll revert

File: zirgen/circuit/verify/wrap_zirgen.cpp [hunk 3]
```diff
@@ -209,17 +217,23 @@ Val CircuitInterfaceZirgen::compute_poly(llvm::ArrayRef<Val> u,
...
                                                     mulMapping.lookup(op.getInner())));
         })
-        .Case<arith::ConstantOp, Zll::PolyOp, ZStruct::SubscriptOp, ZStruct::ArrayOp>(
-            [&](auto op) { builder.clone(*op, mapping); })
-        .Case<Zhlt::ReturnOp>(
-            [&](auto returnOp) { result = totMapping.lookup(returnOp.getValue()); })
+        .Case<arith::ConstantOp, Zll::PolyOp, ZStruct::SubscriptOp, ZStruct::ArrayOp>([&](auto op) {
+          Operation* cloned = builder.clone(*op, mapping);
+          // Return type might be different, since we're likely to be operating on extension field
+          // elements.
```
[the remark this answers is not in the document]
Comment (shkoo) [reply zirgen/circuit/verify/wrap_zirgen.cpp:230]: `zll.const` can give us a base field element, as can `zll.get_global`

File: zirgen/Dialect/Zll/IR/IR.cpp [hunk 4]
```diff
@@ -66,4 +66,29 @@ std::string getLocString(mlir::Location loc) {
...
+                                 op->getAttrDictionary(),
+                                 op->getPropertiesStorage(),
+                                 op->getRegions(),
+                                 newTypes)))
+    return;
+
+  if (mlir::TypeRange(newTypes) != op->getResultTypes()) {
+    for (auto [newType, result] : llvm::zip_equal(newTypes, op->getResults())) {
+      if (llvm::isa<ConstOp>(op))
+        llvm::errs() << "Reinfering " << result << " -> " << newType << "\n";
```
[the remark this answers is not in the document]
Comment (shkoo) [reply zirgen/Dialect/Zll/IR/IR.cpp:83]: Nope, will remove debugging

File: zirgen/circuit/verify/wrap_zirgen.cpp (hunk 1 above)
Comment (shkoo) [reply zirgen/circuit/verify/wrap_zirgen.cpp:258]: Added comments

File: zirgen/bootstrap/src/main.rs [hunk 5]
```diff
@@ -66,6 +66,24 @@ const KECCAK_RUST_OUTPUTS: &[&str] = &[
...
 
+const CALCULATOR_RUST_OUTPUTS: &[&str] = &[
+    "taps.rs",
+    "info.rs",
+    "poly_ext.rs",
+    "defs.rs.inc",
+    "types.rs.inc",
+    "layout.rs.inc",
+    "steps.rs.inc",
+    "validity.rs.inc",
```
[the remark this answers is not in the document]
Comment (shkoo) [reply zirgen/bootstrap/src/main.rs:77]: Commented

File: zirgen/circuit/verify/wrap_zirgen.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jacobdweightman) [reply zirgen/circuit/verify/wrap_zirgen.cpp:234]: Ah, right. All the register values are ext elements here, but constants are still base field. Got it.

**Pull request merged**